optimizely/event/event_factory.py

```diff
@@ -0,0 +1,181 @@
+          continue
+        else:
+          result[k] = v
+      return result
+
+    if not isinstance(user_events, list):
+      user_events = [user_events]
+
+    visitors = []
+
+    for user_event in user_events:
+      visitors.append(cls._create_visitor(user_event, logger))
+      user_context = user_event.event_context
+
+      event_batch = EventBatch(
+        user_context.account_id,
+        user_context.project_id,
+        user_context.revision,
+        user_context.client_name,
+        user_context.client_version,
+        user_context.anonymize_ip,
+        True
+      )
+
+    if len([x for x in visitors if x is not None]) == 0:
+      return None
+
+    event_batch.visitors = visitors
+
+    event_params = json.loads(
+        json.dumps(event_batch.__dict__, default=lambda o: o.__dict__),
+        object_pairs_hook=_dict_clean
+      )
+
+    return LogEvent(cls.EVENT_ENDPOINT, event_params, cls.HTTP_VERB, cls.HTTP_HEADERS)
+
+  @classmethod
+  def _create_visitor(cls, user_event, logger):
+    if not user_event:
+      return None
+
+    if isinstance(user_event, ImpressionEvent):
+      decision = Decision(
+        user_event.experiment.layerId if user_event.experiment else None,
+        user_event.experiment.id if user_event.experiment else None,
+        user_event.variation.id if user_event.variation else None
+      )
+
+      snapshot_event = SnapshotEvent(
+        user_event.experiment.layerId if user_event.experiment else None,
+        user_event.uuid,
+        cls.ACTIVATE_EVENT_KEY,
+        user_event.timestamp
+      )
+
+      snapshot = Snapshot([snapshot_event], [decision])
+
+      visitor = Visitor([snapshot], user_event.visitor_attributes, user_event.user_id)
+
+      return visitor
+
+    elif isinstance(user_event, ConversionEvent):
+      revenue = event_tag_utils.get_revenue_value(user_event.event_tags)
+      value = event_tag_utils.get_numeric_value(user_event.event_tags, logger)
+
+      snapshot_event = SnapshotEvent(
+        user_event.event.id if user_event.event else None,
+        user_event.uuid,
+        user_event.event.key if user_event.event else None,
+        user_event.timestamp,
+        revenue,
+        value,
+        user_event.event_tags
+      )
+
+      snapshot = Snapshot([snapshot_event])
+
+      visitor = Visitor([snapshot], user_event.visitor_attributes, user_event.user_id)
+
+      return visitor
+
+    else:
+      # include log message for invalid event type
+      return
+
+  @staticmethod
+  def build_attribute_list(attributes, project_config):
+    """ Create Vistor Attribute List.
+
+    Args:
+      attributes: Dict representing user attributes and values which need to be recorded.
+      project_config: Instance of ProjectConfig.
+
+    Returns:
+      List consisting of valid attributes for the user. Empty otherwise.
+    """
+
+    if project_config is None:
+      return None
+
+    attributes_list = []
+
+    if isinstance(attributes, dict):
```

Comment:
When is `attributes` not a `dict`? How could it happen?

Reply:
Optimizely client may call an API without passing attributes in which case attributes will be `None`. Updated docstring accordingly.